Webcrypto, v2 paste format, zlib compression

elrido opened the pull request:
This PR fixes #28, fixes #49, fixes #74, fixes #193, fixes #260, fixes #328.

## Changes
* Removes the following libraries:
  * SJCL.js -> replaced by browser WebCrypto API
  * Base64.js -> replaced by browser atob / btoa functions
  * rawdeflate.js -> replaced by zlib.c, crosscompiled to a WASM module (note that rawinflate.js is still present, but only used for v1 paste decryption)
* Frontend able to decrypt v1 and v2 pastes and comments and URLs, newly created pastes and comments are v2 only
* Backend able to send v1 and v2 (pre-v1 were already converted on the fly to v1 API) pastes and comments, but accepts only v2 pastes and comments to store.
* Refactored backends data passing methods to all work with PHP arrays instead of PHP objects in order to avoid duplicate conversions from and to JSON strings when passing pastes/comments between controller, data model and data storage.
* Not changed: The database table structure remains unchanged. The default filesystem storage always stores the JSON as received by the API version at the time of storage.

## ToDo
* [x] Test mixed formats - pastes created on one version with comments from v1 as well as v2
* [x] Import unit test for #260 and check if it is solved by zlib as well
* [x] Have peers review the changes, implement suggested improvements
* [x] Review impact on code coverage, supply additional unit tests if severe reduction of coverage discovered

elrido commented:
### Browser compatibility

TL;DR: Except for IE, all major browsers support the new features as of September 2017.

Apart from the WebCrypto API, this branch also makes use of some other recent features that break backwards compatibility. For reference, here is a list of features and in which version these got supported by the major browsers:

| Browser/Feature | async/await | Web Crypto API | WebAssembly | ECMAScript 2015 (ES6) | first version to support all features | Release date |
|-----------------|-------------|----------------|-------------|-----------------------|---------------------------------------|--------------|
| Chrome          | 55          | 37             | 57          | 51                    | 57                                    | 2017-03-09   |
| Edge            | 15          | 12             | 16          | 15                    | 16                                    | 2017-09-26   |
| Firefox         | 52          | 34             | 52          | 54                    | 54                                    | 2017-06-13   |
| IE              | no          | 11             | no          | 11                    | none                                  | none         |
| Opera           | 42          | 24             | 44          | 38                    | 44                                    | 2017-03-21   |
| Safari          | 10.1        | 11             | 11          | 10                    | 11                                    | 2017-09-19   |

elrido commented:
I have found a regression while testing: The nickname in comments is now part of the encrypted content. Previously an icon/vizhash was only generated when a nickname was set. Since this is now not known to the server, the current logic generates an icon for every comment. The client then displays the comments for anonymous comments with the headline of (same as if there was a nickname):
[icon] _Anonymous_ [localized date string]
instead of just
_Anonymous_ [localized date string]

While we could just hide the icon if it's an anon comment to preserve the old behaviour, I think that would be misleading. I would therefore suggest we document this change of behaviour in the release notes and let the server admins decide if they want to turn it off for everyone, if it bothers them.

Both the vizhash and identicon libraries are designed and used as to not to leak the IP, but it is in their nature that the commentators are less anonymous then without these icons. If that is problematic for a particular instance, the icons should be disabled, like we already [document in the configuration wiki](https://github.com/PrivateBin/PrivateBin/wiki/Configuration#icon).

rugk commented:
> While we could just hide the icon if it's an anon comment to preserve the old behaviour, I think that would be misleading.

:+1: 
Despite of disabling vizhash (which is AFAIK already possible for admins) I also see no reason for _not_ having an avatar for an anonymous comment. (It actually makes sense to identify these comments again.) IMHO, we don't need a setting for this. Let's not overcomplicate it with not needed settings…

elrido commented:
I have now done further testing on the display of existing version 1 pastes with the new logic in the browser as well as the support of mixed version content (paste in v1, comments in v2). Found and fixed some issues with that.

Also solved a regression from base58 that would have made the new pastes occasionally undecryptable as it's encoding strips NULL bytes from the start of the key.

Have also implemented patches suggested by StyleCI and Scrutinizer.

As indicated in #193, I have set up a test server for further review of the new implementation and testing in more browsers then just the latest Firefox and Chrome:
* Test instance: https://zerobin-test.dssr.ch/
* Version 1 paste with version 1 & 2 comments: https://zerobin-test.dssr.ch/?303f5916ec977d79#7WjeUFTClr6bSxxIt4yKXxY/lSij995DowbY319Lws4=
* Version 2 paste with the two formerly failing cases (#260 & #328), that now work using zlib: https://zerobin-test.dssr.ch/?3f711142eea28794#CrjvhH5TnN9YA89MkF4ZWnFuLQYXq82AgVmJHSFEff68

rugk commented on js/privatebin.js:
```diff
@@ -4245,27 +4245,34 @@ jQuery.PrivateBin = (function($, RawDeflate) {
-            } else {
+            let format = '',
+                text   = '';
+            if (paste.hasOwnProperty('ct')) {
```
Refactor to some function like `getPasteVersion` or maybe better `isVersion2Paste` or so?

elrido replied:
Sure

elrido replied:
Given this code smell, I have extracted several of such version specific pieces of logic into generic Paste and Comment classes. The version can now be checked via the `v` attribute that gets set to 1 by default and version 2 pastes will overwrite. See c4b84b2b6bd8

cryptomilk commented:
Just looked at the test instance. I really like the features and new dark theme! Is there already documentation for the v2 paste format?

elrido commented:
To clarify: the test instance used the dark bootstrap theme and has the warning banner, the upload and url shortener features enabled, but the URL shortner requires authentication to work and is usable only by myself for testing. All of these features did not change and are already available in the many of the previous versions, just not enabled by default.

Yes, we did discuss and document the v2 API extensively in the [main issue](https://github.com/PrivateBin/PrivateBin/issues/28#issuecomment-432415707).  We still use JSON-LD, so you can just look at the *.jsonld files found in the js folder, same as before:
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/types.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/paste.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/pastemeta.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/comment.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/commentmeta.jsonld

The [API wiki page](https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format#format-version-2-privatebin--13) is slightly out of date regarding the JSON structure, but already explains correctly the encryption.

I'll gladly have a look at your python client and could create a pull request with changes for this. We could also publish it as a project directly in the PrivateBin organization as a reference. I'd later also have to look at the nodejs client I had created and update it accordingly.

cryptomilk commented:
Then I will look into adding support for v2 in the next days: https://git.cryptomilk.org/projects/cpaste.git/

rugk commented on cfg/conf.sample.php:
```diff
@@ -70,7 +70,7 @@
 ; Note: If you use a bootstrap theme, you can remove the allow-popups from the sandbox restrictions.
 ; By default this disallows to load images from third-party servers, e.g. when they are embedded in pastes. If you wish to allow that, you can adjust the policy here. See https://github.com/PrivateBin/PrivateBin/wiki/FAQ#why-does-not-it-load-embedded-images for details.
-; cspheader = "default-src 'none'; manifest-src 'self'; connect-src *; script-src 'self'; style-src 'self'; font-src 'self'; img-src 'self' data:; media-src data:; object-src data:; Referrer-Policy: 'no-referrer'; sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-modals"
+; cspheader = "default-src 'none'; manifest-src 'self'; connect-src *; script-src 'self' 'unsafe-eval'; style-src 'self'; font-src 'self'; img-src 'self' data:; media-src data:; object-src data:; Referrer-Policy: 'no-referrer'; sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-modals"
```
Arrgh, why eval? For what thing is that needed?
As the name says, one really should not be using it…

rugk replied:
Ah okay, read your commit message. So seems to be an issue with Crome… arrgh… https://github.com/WebAssembly/content-security-policy/issues/7

elrido replied:
I hadn't noticed developing, as I had CSP and HTTPS turned off in the Janitor container

cryptomilk commented:
Congratulations! I will look into the Python client as soon as I have time. Probably after Pfingsten.

r4sas commented:
I think firstly better update API page, add version-depended [en/de]coding references.

elrido commented:
@r4sas, thanks for reminding me. I did go through both the [API](https://github.com/PrivateBin/PrivateBin/wiki/API) and [encryption format](https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format) pages one more time and updated them with the latest changes.

elrido commented:
@r4sas & @cryptomilk - out of curiosity I have looked at one of your python clients and [got it working for the paste creation & deletion](https://github.com/PrivateBin/PrivateBin-Python-Cli). I will now also change the decryption in it and then raise a merge request back to the original repository. Optionally we can also maintain such a client in the main projects organization, if you like.

cryptomilk commented:
I'm fine with maintaining an upstream client and helping out. However to make it easy for cpaste users I will still provide a singe python file as a cpaste client :-)

I'm on the way to the Libre Graphics Meeting right now. Next week is SambaXP conference so I don't really have time to look into privatebin clients right now. However added really awesome stuff to PrivateBin!

r4sas commented:
> I will now also change the decryption in it

According to latest commits, you want drop v1 format support fully?

elrido commented:
That is a good point. Apart from sending, folks may want to fetch with your client as well. Also they may not know what version a certain instance is. Thanks to our JSON-LD files we should be able to implement API version detection and then send the according version.

How about I conclude the "get" / decryption part of the client for v2 format, then it is easier to compare what has changed overall. If I add some version detection to the PrivateBin class in api.py, we could maybe have two classes that inherit some base functionality and have the same API for the encryption/paste format and then the logic in the actions.py would just use the class that matches the servers API for creating/parsing the paste?

r4sas commented:
About creation - I think removing v1 is not problem. But about "get", in json we have `"v": 2` field, which can be used to detect which class use for decryption.
